Hi @zhemao (CC @bluepeppers) 

Thanks very much -- but did you test the patch?  After running 'setup.py develop' to pick up funktown, it still fails the unit tests for me:

(django-pydev-venv)mac 05:08:54 (zhemao) >  nosetests

# ..EE..E.

## ERROR: pyclojure.test_lisp.test_core

Traceback (most recent call last):
  File "/Users/jacobsen/django-pydev-venv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jacobsen/Programming/Lisp/Clojure/pyclojure/pyclojure/test_lisp.py", line 65, in test_core
    Map()["1"] = 2
  File "/Users/jacobsen/Programming/Lisp/Clojure/pyclojure/pyclojure/core.py", line 22, in **setitem**
    self.__dict[name] = value
TypeError: 'ImmutableDict' object does not support item assignment

# 

## ERROR: pyclojure.test_lisp.test_eval

Traceback (most recent call last):
  File "/Users/jacobsen/django-pydev-venv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jacobsen/Programming/Lisp/Clojure/pyclojure/pyclojure/test_lisp.py", line 85, in test_eval
    m[1] = 2
  File "/Users/jacobsen/Programming/Lisp/Clojure/pyclojure/pyclojure/core.py", line 22, in **setitem**
    self.__dict[name] = value
TypeError: 'ImmutableDict' object does not support item assignment

# 

## ERROR: pyclojure.test_lisp.test_to_string

Traceback (most recent call last):
  File "/Users/jacobsen/django-pydev-venv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jacobsen/Programming/Lisp/Clojure/pyclojure/pyclojure/test_lisp.py", line 161, in test_to_string
    assert tostring(parse("{1 2}")) == "{1 2}"
  File "/Users/jacobsen/django-pydev-venv/lib/python2.6/site-packages/ply/yacc.py", line 265, in parse
    return self.parseopt_notrack(input,lexer,debug,tracking,tokenfunc)
  File "/Users/jacobsen/django-pydev-venv/lib/python2.6/site-packages/ply/yacc.py", line 971, in parseopt_notrack
    p.callable(pslice)
  File "/Users/jacobsen/Programming/Lisp/Clojure/pyclojure/pyclojure/parser.py", line 107, in p_map
    p[0] = make_map(p[2])
  File "/Users/jacobsen/Programming/Lisp/Clojure/pyclojure/pyclojure/parser.py", line 31, in make_map
    m[k] = v
  File "/Users/jacobsen/Programming/Lisp/Clojure/pyclojure/pyclojure/core.py", line 22, in **setitem**
    self.__dict[name] = value
TypeError: 'ImmutableDict' object does not support item assignment

---

Ran 8 tests in 0.183s

FAILED (errors=3)
(django-pydev-venv)mac 05:08:56 (zhemao) >

--- a/pyclojure/core.py
+++ b/pyclojure/core.py
@@ -18,9 +18,6 @@ class Map(ComparableExpr):
     def __getitem__(self, name):
         return self.__dict[name]
 
-    def __setitem__(self, name, value):
-        self.__dict[name] = value
-
     def __repr__(self):
         return 'MAP(%s)' % (self.__dict)
 
@@ -44,7 +41,7 @@ class Atom(ComparableExpr):
         return "ATOM(%s)" % (self.__name)
 
 
-class ComparableList(ComparableExpr):
+class List(ComparableExpr):
     def __init__(self, *args):
         self.__contents = []
         for arg in args:
@@ -56,10 +53,6 @@ class ComparableList(ComparableExpr):
     def __repr__(self):
         return "%s(%s)" % (self.__class__.__name__.upper(),
                            ','.join([str(el) for el in self.__contents]))
-
-
-class List(ComparableList):
-    pass
 
 
 class Vector(ComparableExpr):

--- a/pyclojure/parser.py
+++ b/pyclojure/parser.py
@@ -25,11 +25,11 @@ class LispLogger(yacc.PlyLogger):
             super(type(self), self).debug(*args, **kwargs)
 
 def make_map(args):
-    m = Map()
+    m = {}
     kvlist = [(args[i], args[i+1]) for i in range(0, len(args), 2)]
     for k, v in kvlist:
         m[k] = v
-    return m
+    return Map(m)
 
 class PyClojureParse(object):
     def build(self):

--- a/pyclojure/test_lisp.py
+++ b/pyclojure/test_lisp.py
@@ -62,7 +62,6 @@ def test_core():
     assert Map(x=1) != Map(x=2)
     assert Map(x=1) != Map(x=1, a=3)
     assert Map(x=1)["x"] == 1
-    Map()["1"] = 2
 
 
 def evalparser():
@@ -81,10 +80,9 @@ def test_eval():
     assert evalparse("[]") == Vector()
     assert evalparse("[1 2 3]") == Vector(1, 2, 3)
     assert evalparse("{}") == Map()
-    m = Map()
-    m[1] = 2
+    m = Map({1:2})
     assert evalparse("{1 2}") == m
-    m[3] = 4
+    m = Map({1:2, 3:4})
     assert evalparse("{1 2, 3 4}") == m
 
     try:

